modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -344,44 +409,69 @@
+        done-chan  (a/promise-chan)
+        res-fut    (future
+                     (log/trace "BigQuery exec thread sending query job")
+                     (try
+                       (let [result (.query client (.build request) (u/varargs BigQuery$JobOption))]
+                         (log/trace "BigQuery request finished successfully; sending the query to the channel")
+                         (let [pushed? (a/>!! done-chan [:done result])]
+                           (log/tracef "BigQuery channel send complete: %s" (pr-str pushed?)))
+                         nil)
+                       (catch Throwable t
+                         (a/>!! done-chan [:error t]))))]
+    (loop [[res chan] (a/alts!! (cond-> [done-chan]
+                                  cancel-chan (conj cancel-chan))
+                                :priority true)]
+      (if-let [[tag value] (when (= chan done-chan)
+                             res)]
+        ;; done-chan fired first, with either an :error or :done
+        (case tag
+          :done  value
+          :error (condp instance? value
+                   java.util.concurrent.CancellationException
+                   (throw (ex-info (tru "Query cancelled")
+                                   {:sql sql :parameters parameters ::cancelled? true}))
+                   BigQueryException
+                   (let [bqe ^BigQueryException value]
+                     (if (.isRetryable bqe)
+                       (throw (ex-info (tru "BigQueryException executing query")
+                                       {:retryable? (.isRetryable bqe)
+                                        :sql        sql
+                                        :parameters parameters}
+                                       bqe))
+                       (throw-invalid-query bqe sql parameters)))
+                   Throwable
+                   (throw-invalid-query value sql parameters)))
+
+        ;; cancel-chan fired first, it either got closed (nil) or received a real cancel message.
+        (if res
+          ;; Real cancel
+          (do (log/trace "BigQuery cancellation channel fired - cancelling the future if it's not already done")
+              (reset! cancel-requested? true)
+              (when-not (or (future-cancelled? res-fut) (future-done? res-fut))
+                (log/trace "BigQuery initial call still running, cancelling the future")
+                (future-cancel res-fut))
+              ;; And throw about the query being cancelled.
+              (throw (ex-info (tru "Query cancelled")
+                              {:sql sql :parameters parameters ::cancelled? true})))
+
+          ;; Just closed, keep waiting for the `done-chan` to fire.
+          ;; I think this is a race condition for single-row queries?
+          (do (log/trace "BigQuery cancel-chan was closed before query was done - still waiting for query result")
```

Comment:
I'm wondering if this really makes sense, cancel chan being closed with no message before the result comes in is certainly not something we would expect to happen normally -- it would seem to suggest a request cancelation wasn't getting detected correctly